### DISCUSSION TITLE
Setting up failover cluster with external replication

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#14150] [Q&A]
Hi!

I'm trying to setup a secondary cluster for failover, using the External Replication task.

However, I noticed that the compare exchange values are not transferred.

Is this a problem in my configuration or?

**Comment by garayx** [comment]
That's by design, cluster-level features are not replicated, see: 
https://ravendb.net/docs/article-page/5.3/csharp/server/ongoing-tasks/external-replication#general-information-about-external-replication-task

**Reply by georgiosd** [reply]
Hm.... so how can failover work in this scenario? If I switch from cluster A to cluster B when A goes down but it's missing all the cmp xchg data, the applications are not going to work.

**Comment by ayende** [comment]
A few things about this:

* A secondary cluster would require you to do manual failover between the clusters. There is no failover between clusters. Each are independent.
* A cmp xchg isn't replicated between clusters because the issue is with consistency. 

More to the point, the issue is the consistency between clusters. See also here: https://ayende.com/blog/196769-B/data-ownership-in-a-distributed-system

Consider the simplest scenario, using cmp xchg  to ensure unique email.
If you are working in two clusters, what is the _meaning_ of replication of the cmp xchng values between them?
If you can modify the cmp xchg in both... there is no consistency boundary.

Note that you _can_ send the _data_. You can setup ETL process that would read using `cmpxchg` and send them (as documents) to the other side. But that is explicitly not about creating cmp xchg in the other cluster.

**Reply by georgiosd** [reply]
Yes, I see the predicament you're presenting, if both clusters can write to each other.

I'm not seeing the solution however.

If the cmp xchg are not carried over, the application must have some sort of logic that disables cluster-wide ops if using the secondary cluster, otherwise reserved emails will become available again.

I could imagine a one way replication only (Primary => Secondary) that includes cmp xchg, then I *can* (manually) fail over to the secondary cluster that I then have to switch back manually again. Where this would break down is if when the primary cluster comes back up, there are changes that it didn't finish sending and sends them again, where consistency would break.

Is there another solution here that I'm missing on having two clusters that can facilitate manual failover?

**Reply by georgiosd** [reply]
Just bumping this - wondering what I'm missing here on a setup  that is able to have failover capabilities, even if manual.

**Reply by garayx** [reply]
Another option to transfer the compare exchanges to the 2nd cluster, is exporting the compare exchanges to a dump file on the 1st cluster, then importing them on the 2nd.
https://ravendb.net/docs/article-page/5.1/csharp/studio/database/tasks/export-database#export-options

**Reply by ayende** [reply]
Alternatively, you can use cluster wide transactions, which will "work".
The documents will move over, and you'll need to manage how to ensure that you aren't having concurrent cluster wide transactions on both sides in a separate channel. But that would carry over properly and you won't need to think about this most times.

**Reply by ayende** [reply]
Might be that a better question here is what is the _business_ scenario at hand?

**Reply by georgiosd** [reply]
The business scenario is that we are hosting multiple client applications on a single RavenDB cluster that is hosted within the same datacenter. Most functionality is provided with single-node writes but there are some use cases that depend on cmp xchg (e.g. user registration).

Assuming a second, separate, cluster located in another data center that remains idle until the first zone fails, we want to entertain how we could continue providing full functionality to the users with near-zero downtime.

I do understand that we will have to take some sort of action to actually make the switch from one cluster to the other, and right now I am making the assumption it will be a script responding to the load balancer signaling all backends being down, or the load balancer being unreachable itself.

I hope this is what you meant by business scenario :)

**Reply by ayende** [reply]
Okay, yes, that make sense.
Given that scenario, you have a simpler mode of operations.
Your primary cluster is the one that allows creation of users.
Your secondary cluster is *not* allowing that unless you explicitly set it (as part of failover).

Replication & everything will work. And cluster wide transactions will operate normally and safely in this situation.

**Reply by georgiosd** [reply]
When you say that the secondary cluster is _not_ allowing that, how do you mean? Do you envisage the app being aware that it's running on the secondary cluster and blocking operations that depend on cluster-wide ops?

I would have to set up the cmp xchg ETL process for each database as you mentioned earlier, right?

Any ideas on how to block incoming replication/ETL from the secondary cluster? I'm worried about the primary coming online and sending leftovers and something blowing up.

**Reply by ayende** [reply]
Yes, the idea is that for the secondary cluster to accept those operations, you need a manual switch.
A common way to do that is to have a document that would say `{Owner: '1st Cluster'}` and check that as part of the transaction.

For the cmp/xchg - what are you doing there? Can you do that as a cluster wide transaction, in which case the documents will be replicated for you?

**Reply by georgiosd** [reply]
You mean to always use compare exchange from a cluster wide operation vs using the store operations? Yes I do that most of the time.

I'm not sure I understand though - are you saying that external replication will, in this case, copy over the compare exchange keys?

So that we dont get lost in language, lets give it an example of a create user:
A. users/1-A (document)
B. emails/foo@bar.com => users/1-A (cmp xchg)

**Reply by ayende** [reply]
Don't do cmp xchg at all.
Instead, use the following documents in a cluster wide transaction:

* `users/1-A` - `{Email: 'foo@bar.com'}`
* `emails/foo@bar.com` - `{id: 'users/1-a'}`

Note that those are _documents.

Your code will always ensure that any modification to `emails/` is done in a cluster wide tranaction.


That gives you:

* Ensure safety in a cluster wide distributed transaction
* Replicate the _documents_ to the other side

Note that it won't give you cross cluster safety, but it handle everything else.

**Reply by georgiosd** [reply]
![](https://user-images.githubusercontent.com/1672054/167999103-18ad513c-b446-4662-b156-4363a69b264b.gif)

IMO that advice needs to be added here! 
https://ravendb.net/docs/article-page/5.3/csharp/client-api/operations/compare-exchange/overview#example-i---email-address-reservation

Thanks Oren - what use cases would you recommend the direct use of compare exchange?

**Reply by ayende** [reply]
Good point: https://issues.hibernatingrhinos.com/issue/RDoc-2183

I don't think there are many such scenarios, to be honest. 
The one that pops to mind is if you need to do things like expiration directly.

**Comment by georgiosd** [comment]
Thank you. 

I would gladly contribute to the docs more but the doc system needs quite a bit of time to do right for someone who is not familiar.

Would it be a fair compromise to offer the text and then have someone on your side do the actual edits in the source code?

**Reply by ppekrol** [reply]
@georgiosd You can paste your proposal to the issue that @ayende sent. We will cover all technicalities then.
